README.md
# canvas-timeliner

![Preview](./snapshot.jpg)

## Install

```
yarn add @tomieric/canvas-timeliner
```

## Usage

```
import Ruler from '@tomieric/canvas-timeliner'

const ruler = new Rule(options?)
```


## Options

```
export interface IOptions {
  container: HTMLElement;
  position: string;
  textPos: number;
  textAlign: string;
  gap: number;
  width: number;
  height: number;
  lineOffset: number;
  gapHeight: number;
  gapColor: number;
  lineColor: string;
  lineWidth: number;
  lineHeight: number;
  interval: number;
  textSize: string;
  textStyle: string;
  autoSize: boolean;
  formatMarkText: (mark: number | string) => string;
}
```

### default options

```
const DEFAULT_OPTIONS: object = {
  height: 100,
  textSize: 10,
  interval: 10,
  gap: 10,
  textPos: 5,
  gapHeight: 20,
  lineHeight: 10,
  lineWidth: 1,
  lineColor: 'rgb(0, 0, 0)',
  textStyle: '#000000'
}
```


## Method

  * render(options?) redraw to canvas

```
const instance = new CanvasRuler(options?)
// resize ruler
instance.render({
  width: 3000,
  interval: 10,
  gap: 20
})
```

## Development

```bash
yarn dev
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0-alpha.1](https://github.com/tomieric/canvas-timeliner/compare/v1.1.0-alpha.0...v1.1.0-alpha.1) (2020-03-09)


### Features

* **canvas-timeliner:** add textAlign property ([3a041d8](https://github.com/tomieric/canvas-timeliner/commit/3a041d8c64884b2d05f0643fa4ffbc5dcd480216))
* **canvas-timeliner:** update ([c20587b](https://github.com/tomieric/canvas-timeliner/commit/c20587b589edaf888bf2209a4266a187b3526355))
* **ruler:** 修改 render 函数 ([ae0d114](https://github.com/tomieric/canvas-timeliner/commit/ae0d114508003d7cc750174c34b4e36a4921af87))


### Bug Fixes

* **options:** add options ([ba1fe1c](https://github.com/tomieric/canvas-timeliner/commit/ba1fe1cccc74231de2763f0ebfb6c8fdd8ea9b59))

## 1.1.0-alpha.0 (2020-03-02)


### Features

* **canvas-timeliner:** 增加左右类型 ([96e3cfd](https://github.com/tomieric/canvas-timeliner/commit/96e3cfdadcc3b008101ef78b6e9de4ea3c8fbe37))

### [1.0.2](https://github.com/tomieric/canvas-timeliner/compare/v1.0.1...v1.0.2) (2020-03-01)


example/index.js
import './style.css'
import TimeLiner from '../dist/lib'

const app = document.getElementById('app')
const darkApp = document.getElementById('app-dark')
const leftApp = document.getElementById('app-left')
const rightApp = document.getElementById('app-right')
const points = document.getElementById('points')

const padLeft = n => n >= 10 ? n : `0${n}` 

console.log(padLeft(23))

const gap = 24
const interval = 10
const timeliner = new TimeLiner({
  container: app,
  gap,
  formatMarkText (frame) {
    const m = frame >= gap ? Math.floor(frame / gap) : 0
    const f = frame % 24
    return `00:0${m}:${padLeft(f)}`
  }
})

const timeLiner2 = new TimeLiner({
  container: darkApp,
  gap: 24,
  position: 'bottom',
  textAlign: 'center',
  lineHeight: 6,
  gapHeight: 6,
  interval: 10,
  lineColor: '#999999', // rgb(51, 51, 51)
  gapColor: '#eeeeee', // rgb(102, 102, 102)
  textStyle: '#EEEEEE',
  formatMarkText (frame) {
    const m = frame >= gap ? Math.floor(frame / gap) : 0
    const f = frame % 24
    return `00:0${m}:${padLeft(f)}`
  }
})

const leftTimeLine = new TimeLiner({
  container: leftApp,
  gap,
  position: 'left',
  formatMarkText (frame) {
    const m = frame >= gap ? Math.floor(frame / gap) : 0
    const f = frame % 24
    return `00:0${m}:${padLeft(f)}`
  }
})

const rightTimeLine = new TimeLiner({
  container: rightApp,
  gap,
  position: 'right',
  formatMarkText (frame) {
    const m = frame >= gap ? Math.floor(frame / gap) : 0
    const f = frame % 24
    return `00:0${m}:${padLeft(f)}`
  }
})

points.addEventListener('change', e => {
  const val = interval * e.target.value
  console.log(gap, val, gap * val)
  timeliner.draw({
    interval: val
  })
  timeLiner2.draw({
    interval: val,
    gap: Math.ceil(gap * (val / 10))
  })
  leftTimeLine.draw({
    interval: val
  })
  rightTimeLine.draw({
    interval: val
  })
})

rollup.config.js
import typescript from 'rollup-plugin-typescript'
import commonjs from 'rollup-plugin-commonjs'

const pkg = require('./package.json')
const name = pkg.name.split('/').pop()
const version = pkg.version

var banner =
`/*!
 * ${pkg.name} ${version} (https://github.com/tomieric/canvas-timeliner)
 * API https://github.com/tomieric/canvas-timeliner/blob/master/README.md
 * Copyright 2020-preset tommyshao All Rights Reserved
 * Licensed under MIT (https://gitlab.com/tomieric/canvas-timeliner/blob/master/LICENSE)
 */
`

const config = ['cjs', 'es'].map(format => ({
  input: './src/index.ts',
  output: {
    name,
    file: `dist/${format === 'es' ? 'esm' : 'lib'}/index.js`,
    format,
    banner
  },
  plugins: [
    typescript({lib: ["es5", "es6", "dom"], target: "es5"}),
    commonjs({extensions: ['.js', '.ts']})
  ]
}))

export default config


example.b7588b45.js
parcelRequire=function(e,r,t,n){var i,o="function"==typeof parcelRequire&&parcelRequire,u="function"==typeof require&&require;function f(t,n){if(!r[t]){if(!e[t]){var i="function"==typeof parcelRequire&&parcelRequire;if(!n&&i)return i(t,!0);if(o)return o(t,!0);if(u&&"string"==typeof t)return u(t);var c=new Error("Cannot find module '"+t+"'");throw c.code="MODULE_NOT_FOUND",c}p.resolve=function(r){return e[t][1][r]||r},p.cache={};var l=r[t]=new f.Module(t);e[t][0].call(l.exports,p,l,l.exports,this)}return r[t].exports;function p(e){return f(p.resolve(e))}}f.isParcelRequire=!0,f.Module=function(e){this.id=e,this.bundle=f,this.exports={}},f.modules=e,f.cache=r,f.parent=o,f.register=function(r,t){e[r]=[function(e,r){r.exports=t},{}]};for(var c=0;c<t.length;c++)try{f(t[c])}catch(e){i||(i=e)}if(t.length){var l=f(t[t.length-1]);"object"==typeof exports&&"undefined"!=typeof module?module.exports=l:"function"==typeof define&&define.amd?define(function(){return l}):n&&(this[n]=l)}if(parcelRequire=f,i)throw i;return f}({"iMte":[function(require,module,exports) {

},{}],"GCpX":[function(require,module,exports) {
"use strict";var t={height:100,textSize:10,interval:10,gap:10,textPos:5,gapHeight:20,lineHeight:10,lineWidth:1,lineColor:"rgb(0, 0, 0)",textStyle:"#000000"},e=5,i=function(){function i(e){var i=document.createElement("canvas"),s=e.container,n=e.autoSize;s?(this.canvas=i,this.container=s,this.ctx=i.getContext("2d"),this.options=Object.assign({},t,e),this.init(),this.render(),n&&this.resizeObserve()):console.warn("container must be to set")}return i.prototype.init=function(t){var e=t||this.container;e?e.appendChild(this.canvas):console.log("container must be to set")},i.prototype.render=function(t){t&&this.setOption(t),this.setRect(),this.draw()},i.prototype.setRect=function(t){var e=this.canvas,i=this.container,s=this.options,n=s.width,o=s.height,h=i.getBoundingClientRect();e.width=this.width=n||h.width,t||(e.height=this.height=o||h.height)},i.prototype.draw=function(){var t=this.width,e=this.options,i=e.position,s=e.interval;/left|right/.test(i)&&(t=this.height),this.steps=Math.round(t/s),this.ctx.clearRect(0,0,this.width,this.height),this.drawBorderLine(),this.drawMark()},i.prototype.drawBorderLine=function(){var t=this.options.position,e=[0,0,this.width,this.height],i=e[0],s=e[1],n=e[2],o=e[3];switch(t){case"bottom":i=0,s=this.height;break;case"left":n=0;break;case"right":i=this.width;break;case"top":default:s=0,o=0}this.drawLine(i,s,n,o)},i.prototype.drawLine=function(t,e,i,s){var n=this.ctx,o=this.options,h=o.lineColor,r=o.lineWidth;n.beginPath(),n.moveTo(t,e),n.lineTo(i,s),n.strokeStyle=h,n.lineWidth=r,n.stroke()},i.prototype.drawText=function(t,e,i){var s=this.ctx,n=this.options,o=n.textStyle,h=n.textSize;s.font=h,s.fillStyle=o,s.textBaseline="middle",s.fillText(t,e,i)},i.prototype.drawMark=function(){for(var t=this,i=this.ctx,s=this.options,n=s.lineHeight,o=s.gapHeight,h=s.gap,r=s.interval,a=s.formatMarkText,c=s.position,d=s.textSize,p=s.textPos,l=s.textAlign,w=!1,g=!1,x=Number(d),v=1.5*x,u=function(s){var h=[0,0,0,0,0,0],d=h[0],g=h[1],u=h[2],f=h[3],b=h[4],y=h[5],m=a?a(s):s,k=i.measureText(m).width,z=Math.max(r,e);switch(c){case"left":d=0,u=(w?o:n)+.5,g=f=z*s+.5,b=(w?o:n)+p+.5,y=s>0&&"center"===l?g+v/4:g+v/2;break;case"right":d=t.width-(w?o:n)+.5,u=t.width,g=f=z*s+.5,y=s>0&&"center"===l?g+v/4:g+v/2,b=t.width-o-k-p;break;case"bottom":d=u=z*s+.5,g=t.height-(w?o:n)+.5,f=t.height,b=s>0&&"center"===l?d-k/2:d,y=t.height-o+.5-p;break;case"top":default:d=u=z*s+.5,g=0,f=(w?o:n)+.5,b=s>0&&"center"===l?d-k/2:d,y=o+.5+x+p}return{x:d,y:g,x1:u,y1:f,textX:b,textY:y,text:m}},f=0;f<this.steps;f++){g=(w=f>0&&!(f%h))||0===f;var b=u(f),y=b.x,m=b.y,k=b.x1,z=b.y1,O=b.textX,R=b.textY,S=b.text;this.drawLine(y,m,k,z),g&&this.drawText(S,O,R)}},i.prototype.setOption=function(t){this.options=Object.assign(this.options,t)},i.prototype.resizeObserve=function(){var t=this;if(window.ResizeObserver){var e=new window.ResizeObserver(function(e){for(var i=0,s=e;i<s.length;i++){var n=s[i];t.width!==n.contentRect.width&&t.setRect(!0),t.draw()}});e.observe&&e.observe(this.container)}},i}();module.exports=i;
},{}],"Focm":[function(require,module,exports) {
"use strict";require("./style.css");var t=e(require("../dist/lib"));function e(t){return t&&t.__esModule?t:{default:t}}var n=document.getElementById("app"),a=document.getElementById("app-dark"),r=document.getElementById("app-left"),o=document.getElementById("app-right"),c=document.getElementById("points"),i=function(t){return t>=10?t:"0".concat(t)};console.log(i(23));var l=24,u=10,d=new t.default({container:n,gap:l,formatMarkText:function(t){var e=t>=l?Math.floor(t/l):0,n=t%24;return"00:0".concat(e,":").concat(i(n))}}),f=new t.default({container:a,gap:24,position:"bottom",textAlign:"center",lineHeight:6,gapHeight:6,interval:10,lineColor:"#999999",gapColor:"#eeeeee",textStyle:"#EEEEEE",formatMarkText:function(t){var e=t>=l?Math.floor(t/l):0,n=t%24;return"00:0".concat(e,":").concat(i(n))}}),g=new t.default({container:r,gap:l,position:"left",formatMarkText:function(t){var e=t>=l?Math.floor(t/l):0,n=t%24;return"00:0".concat(e,":").concat(i(n))}}),p=new t.default({container:o,gap:l,position:"right",formatMarkText:function(t){var e=t>=l?Math.floor(t/l):0,n=t%24;return"00:0".concat(e,":").concat(i(n))}});c.addEventListener("change",function(t){var e=u*t.target.value;console.log(l,e,l*e),d.draw({interval:e}),f.draw({interval:e,gap:Math.ceil(l*(e/10))}),g.draw({interval:e}),p.draw({interval:e})});
},{"./style.css":"iMte","../dist/lib":"GCpX"}]},{},["Focm"], null)
//# sourceMappingURL=/canvas-timeliner/example.b7588b45.js.map

src/index.ts
export interface IOptions {
  container: HTMLElement; // 容器
  position: string; // 位置
  textPos: number; // 刻度位置
  textAlign: string; // 文字对齐
  gap: number;
  width: number;
  height: number;
  lineOffset: number;
  gapHeight: number;
  gapColor: number;
  lineColor: string; // 标线颜色
  lineWidth: number;
  lineHeight: number;
  interval: number;
  textSize: string;
  textStyle: string;
  autoSize: boolean;
  formatMarkText: (mark: number | string) => string;
}

const DEFAULT_OPTIONS: object = {
  height: 100,
  textSize: 10,
  interval: 10,
  gap: 10,
  textPos: 5,
  gapHeight: 20,
  lineHeight: 10,
  lineWidth: 1,
  lineColor: 'rgb(0, 0, 0)',
  textStyle: '#000000'
}

const MIN_INTERVAL = 5

export default class TimeLiner {
  public canvas: HTMLElement;
  public container: HTMLElement;
  public ctx: CanvasRenderingContext2D;
  public options: IOptions;
  public width: number;
  public height: number;
  public steps: number; // 刻度数

  constructor (options: IOptions) {
    const canvas = document.createElement('canvas')
    const { container, autoSize } = options

    if (!container) {
      console.warn('container must be to set')
      return
    }
    
    this.canvas = canvas
    this.container = container
    this.ctx = canvas.getContext('2d')
    this.options = Object.assign({}, DEFAULT_OPTIONS, options)

    this.init()
    this.render()
    if (autoSize) this.resizeObserve()
  }

  private init (dom?: HTMLElement) {
    const el = (dom || this.container)
    if (el) {
      el.appendChild(this.canvas)
    } else {
      console.log('container must be to set')
    }
  }

  render (options?: IOptions) {
    if (options) this.setOption(options)
    this.setRect()
    this.draw()
  }

  private setRect (isResize?: boolean) {
    const { canvas, container, options } = this
    const { width, height } = options
    const boundRect: ClientRect = container.getBoundingClientRect()

    ;(canvas as any).width = this.width = width || boundRect.width
    if (!isResize) (canvas as any).height = this.height = height || boundRect.height
  }

  private draw () {
    let total = this.width
    const { position, interval } = this.options

    if (/left|right/.test(position)) {
      total = this.height
    }

    this.steps = Math.round(total / interval)
    this.ctx.clearRect(0, 0, this.width, this.height)

    // 绘制长线
    this.drawBorderLine()
    
    // 绘制刻线
    this.drawMark()
  }

  private drawBorderLine () {
    const { position } = this.options
    let [x, y, x1, y1] = [0, 0, this.width, this.height]
    switch (position) {
      case 'bottom':
        x = 0
        y = this.height
        break
      case 'left':
        x1 = 0
        break
      case 'right':
        x = this.width
        break
      case 'top':
      default:
        y = 0
        y1 = 0
        break
    }
    this.drawLine(x, y, x1, y1)
  }

  private drawLine (x: number, y: number, endX: number, endY: number) {
    const ctx = this.ctx
    const { lineColor, lineWidth } = this.options
    ctx.beginPath()
    ctx.moveTo(x, y)
    ctx.lineTo(endX, endY)
    ctx.strokeStyle = lineColor
    ctx.lineWidth = lineWidth
    ctx.stroke()
  }

  private drawText (text: string | number, x: number, y: number) {
    const { ctx, options } = this
    const { textStyle, textSize } = options
    
    ctx.font = textSize
    ctx.fillStyle = textStyle
    ctx.textBaseline = 'middle'
    ctx.fillText(text as string, x, y)
  }

  private drawMark () {
    const { ctx, options } = this
    const { lineHeight, gapHeight, gap, interval, formatMarkText, position, textSize, textPos, textAlign } = options
    let isGap: boolean = false
    let isGapText: boolean = false
    const fontSize = Number(textSize)
    const textHeight = fontSize * 1.5

    // + 0.5 解决 1px 模糊
    const getXY = (num: number) => {
      let [x, y, x1, y1, textX, textY] = [0, 0, 0, 0, 0, 0]
      const text = formatMarkText ? formatMarkText(num) : num
      const textWidth = ctx.measureText(text as string).width
      const intervalN = Math.max(interval, MIN_INTERVAL)

      switch (position) {
        case 'left':
            x = 0
            x1 = (isGap ? gapHeight : lineHeight) + 0.5
            y = y1 = intervalN * num + 0.5
            textX = (isGap ? gapHeight : lineHeight) + textPos + 0.5
            textY = num > 0 && textAlign === 'center' ? y + textHeight / 4 : y + textHeight / 2
          break
        case 'right':
            x = this.width - (isGap ? gapHeight : lineHeight) + 0.5
            x1 = this.width
            y = y1 = intervalN * num + 0.5
            textY = num > 0 && textAlign === 'center' ? y + textHeight / 4 : y + textHeight / 2
            textX = this.width - gapHeight - textWidth - textPos
          break
        case 'bottom':
            x = x1 = intervalN * num + 0.5
            y = this.height - (isGap ? gapHeight : lineHeight) + 0.5
            y1 = this.height
            textX = num > 0 && textAlign === 'center' ? x - textWidth / 2 : x
            textY = this.height - (gapHeight) + 0.5 - textPos
          break;
        case 'top':
        default:
            x = x1 = intervalN * num + 0.5
            y = 0
            y1 = (isGap ? gapHeight : lineHeight) + 0.5
            textX = num > 0 && textAlign === 'center' ? x - textWidth / 2 : x
            textY = gapHeight + 0.5 + fontSize + textPos
          break;
      }

      return { x, y, x1, y1, textX, textY, text }
    }
    
    for (let i = 0; i < this.steps; i++) {
      isGap = i > 0 ? !(i % gap) : !1
      isGapText = isGap || i === 0
      const { x, y, x1, y1, textX, textY, text } = getXY(i)
      this.drawLine(x, y, x1, y1)
      
      isGapText && this.drawText(text, textX, textY)
    }
  }

  setOption (options: IOptions) {
    this.options = Object.assign(this.options, options)
  }

  private resizeObserve() {
    if ((window as any).ResizeObserver) {
      const resizeObserver = new (window as any).ResizeObserver((entries: HTMLElement[]) => {
        for (let entry of entries) {
          if (this.width !== (entry as any).contentRect.width) {
            this.setRect(true)
          }
          this.draw()
        }
      })
      resizeObserver.observe && resizeObserver.observe(this.container)
    }
  }
}
